README.MD

```diff
@@ -1,105 +1,291 @@
+
+Atomicity is achieved using a 2-phase message passing architecture between the chains. Messages are declared on the source chain, and confirmed on the target chain with Merkle Patricia proofs once the source chain is finalized. Once messages are confirmed on the target chain, they can efficiently progressed with a hashlock.
+Messages can also be reverted if they are not yet completed on the target chain.
+
+> ⚠️ Reverting is not yet supported in `mosaic.js` and will be supported in a future version.
+> In the meantime you can revert by accessing the contracts directly.
+
+You can read [the draft of the mosaic whitepaper][mosaic whitepaper] or [the original OpenST whitepaper][openst whitepaper].
+
+## Instructions
+
+### Installation
+
+```bash
+npm install @openstfoundation/mosaic.js web3 web3-eth-accounts
+```
+> ℹ️ Note that `web3` and `web3-eth-accounts` are peer-dependencies.
+
+### Usage
+
+#### Mosaic
 
-# Install Package
 ```js
-npm install @openstfoundation/mosaic.js
-Mosaic = require('@openstfoundation/mosaic.js');
+const Web3 = require('web3');
+const Mosaic = require('@openstfoundation/mosaic.js');
+
+const originWeb3 = new Web3('http://localhost:8546');
+const auxiliaryWeb3 = new Web3('http://localhost:8547');
+
+const originContractAddresses = {
+  EIP20Gateway: '0x0000000000000eip20gatewaycontractaddress',
+  ...,
+};
+const auxiliaryContractAddresses = {
+  EIP20CoGateway: '0x00000000000eip20cogatewaycontractaddress',
+  ...,
+};
+
+const originChain = new Mosaic.Chain(originWeb3, originContractAddresses);
+const auxiliaryChain = new Mosaic.Chain(auxiliaryWeb3, auxiliaryContractAddresses);
+
+const mosaic = new Mosaic(originChain, auxiliaryChain);
 ```
 
-# Abi Bin Provider Usage
+#### Facilitator
+
+Creating a new facilitator:
 ```js
-abiBinProvider = new Mosaic.AbiBinProvider();
-jsonInterface = abiBinProvider.getABI('Organization');
-contractBin = abiBinProvider.getBIN('Organization');
+const facilitator = new Mosaic.Facilitator(mosaic);
+```
+
+Staking:
+```js
+const staker = '0x00000000000000000000stakeraccountaddress';
+const amount = '1000000000000';
+const beneficiary = '0x000000000000000beneficiaryaccountaddress';
+const gasPrice = '5';
+const gasLimit = '150000000';
+const hashLock = '0x00000000000000000000000000000000000000000000000000000000hashlock';
+const txOptions = { from: staker };
+
+facilitator
+  .stake(
+    staker,
+    amount,
+    beneficiary,
+    gasPrice,
+    gasLimit,
+    hashLock,
+    txOptions,
+  )
+  .then(console.log)
+  .catch(console.log);
 ```
 
+Progressing the stake after the state root has been transferred:
```

Comment:
should we add any reference, how this is done?

Reply:
https://github.com/OpenSTFoundation/mosaic-anchor.js

Reply:
we need a more concerted effort to build documentation immediately after beta.1 release